pkgs/checks/lib/src/extensions/async.dart

```diff
@@ -0,0 +1,142 @@
+  /// Expect that the Stream emits a value without first emitting an error.
+  ///
+  /// Returns a Future that completes to a [Check<T>] on the next event emitted
+  /// by the stream.
+  ///
+  /// Fails if the stream emits an error instead of a value, or closes without
+  /// emitting a value.
+  Future<Check<T>> emits() async {
+    return await context.nestAsync<T>('Emits a value', (actual) async {
+      if (!await actual.hasNext) {
+        return Extracted.rejection(
+            actual: 'an empty stream', which: ['did not emit any value']);
+      }
+      try {
+        return Extracted.value(await actual.next);
+      } catch (e) {
+        return Extracted.rejection(
+            actual: 'A stream with error ${literal(e)}',
+            which: ['emittid an error instead of a value']);
+      }
+    });
+  }
+
+  /// Expects that the Stream emits any number of events before emitting an
+  /// event that satisfies [condition].
+  ///
+  /// Returns a Future that completes after the Stream has emitted an even that
+  /// satisfies [condition].
+  ///
+  /// Fails if the Stream emits an error or closes before emitting a matching
+  /// event.
+  Future<void> emitsThrough(void Function(Check<T>) condition) async {
+    await context.expectAsync(
+        () => [
+              'Emits any values then a value that:',
+              ...indent(describe(condition))
+            ], (actual) async {
+      var count = 0;
+      await for (var emitted in actual.rest) {
+        if (softCheck(emitted, condition) == null) {
+          return null;
+        }
+        count++;
+      }
+      return Rejection(
+          actual: 'a stream',
+          which: ['ended after emitting $count elements with none matching']);
+    });
+  }
+
+  /// Expects that the Stream closes without emitting any even that satisfies
+  /// [condition].
+  ///
+  /// Returns a Future that completes after the Stream has closed.
+  ///
+  /// Fails if the Stream emits any even that satisfies [condition].
+  Future<void> neverEmits(void Function(Check<T>) condition) async {
+    await context.expectAsync(
+        () => ['Never emis a value that:', ...indent(describe(condition))],
+        (actual) async {
+      var count = 0;
+      await for (var emitted in actual.rest) {
+        if (softCheck(emitted, condition) == null) {
+          return Rejection(actual: 'a stream', which: [
+            'emitted ${literal(emitted)}',
+            if (count > 0) 'following $count other items'
+          ]);
+        }
+        count++;
+      }
+      return null;
+    });
+  }
+}
+
+extension ChainAsync<T> on Future<Check<T>> {
+  /// Checks the expectations in [condition] against the result of this Future.
+  ///
+  /// Extensions written on [Check] cannot be invoked on [Future<Check>]. This
+  /// method allows adding expectations for the value without awaiting it.
+  ///
+  /// ```dart
+  /// await checkThat(someFuture).completes().that((r) => r.equals('expected'));
```

Comment:
Nitpickinly, "completes" in the language specification covers both evaluating to a value and throwing ("completes by throwing error *e* and stack trace *s*").

We could have

* `Check<Result<T>> completes()`
* `Future<Check<T>> get value`
* `Future<Check<E>> throws<E>()`

to cover all the cases. The `Result` type belongs to `package:async`, but we already have `StreamQueue` specific checks too. (Should those be in a helper library of `package:async` instead? I don't like adding a dependency on another package just to use `package:checks`.)
If you don't like `value`, could it be `completesWith` (short for `completesWithValue`)?

You can do:
```dart
  await checkThat(someFuture).value.then((c) => c.equals(42));
```

If we want to allow `await checkThat(someFuiture).value.equals(42)`, we can also define all the matches on `Check<T>` also on `Future<Check<T>>`, working "below the future" .
So
```dart
extension <T> on Future<Check<T>> {
  Future<void> equals(Object? value) => this.then((c) => c.equals(value));
  Future<Check<R>> that<R>(R check(T value)) => this.then((c) => c.that(check));
  // etc.
}
```
That's a lot of extra declarations just to allow you to directly work on the value below the `Future`
And you can already do that using `then` on the future anyway.


The `.value` getter/`completes()` function converts a `Check<Future<T>>` to a `Future<Check<T>>`, which swaps the monads. That's actually quite an interesting operation. Do we want the opposite, taking a `Future<Check<T>>` and getting a `Check<Future<T>>`? It's probably not particularly useful, since there aren't a lot of things you can do on a future other than wait for it to complete.